File: python/adb_auto_player/ipc/constraint.py
"""ADB Auto Player Input Parameter Constrains Module."""

from typing import Literal, TypedDict


class NumberConstraintDict(TypedDict):
    """Number constraint."""

    type: Literal["number"]
    minimum: float | int
    maximum: float | int
    step: float
    default_value: float | int


class SelectConstraintDict(TypedDict):
    """Select constraint."""

    type: Literal["select"]
    choices: list[str]
    default_value: str


class CheckboxConstraintDict(TypedDict):
    """Checkbox constraint."""

    type: Literal["checkbox"]
    default_value: bool


class MultiCheckboxConstraintDict(TypedDict):
    """Multicheckbox constraint."""

    type: Literal["multicheckbox"]
    choices: list[str]
    default_value: list[str]
    group_alphabetically: bool


class ImageCheckboxConstraintDict(TypedDict):
    """Image checkbox constraint."""

    type: Literal["imagecheckbox"]
    choices: list[str]
    default_value: list[str]
    image_dir_path: str


class MyCustomRoutineConstraintDict(TypedDict):
    """My Custom Routine constraint."""

    type: Literal["MyCustomRoutine"]
    choices: list[str]
    default_value: list[str]


class TextConstraintDict(TypedDict):
    """Text constraint."""

    type: Literal["text"]
    regex: str
    title: str
    default_value: str


ConstraintType = (
    NumberConstraintDict
    | SelectConstraintDict
    | CheckboxConstraintDict
    | MultiCheckboxConstraintDict
    | ImageCheckboxConstraintDict
    | TextConstraintDict
    | MyCustomRoutineConstraintDict
)

File: frontend/src/lib/types/constraint.ts
interface BaseConstraint {
  type: string;
}

interface NumberConstraint extends BaseConstraint {
  type: "number";
  step: number;
  minimum: number;
  maximum: number;
  default_value: number;
}

interface CheckboxConstraint extends BaseConstraint {
  type: "checkbox";
  default_value: boolean;
}

interface MultiCheckboxConstraint extends BaseConstraint {
  type: "multicheckbox";
  choices: string[];
  default_value: string[];
  group_alphabetically: boolean;
}

interface ImageCheckboxConstraint extends BaseConstraint {
  type: "imagecheckbox";
  choices: string[];
  default_value: string[];
  image_dir_path: string;
}

interface SelectConstraint extends BaseConstraint {
  type: "select";
  choices: string[];
  default_value: string;
}

interface MyCustomRoutineConstraint extends BaseConstraint {
  type: "MyCustomRoutine";
  choices: string[];
  default_value: string[];
}

interface TextConstraint extends BaseConstraint {
  type: "text";
  regex: string;
  title: string;
  default_value: string;
}

type ConstraintTypeMap = {
  number: NumberConstraint;
  checkbox: CheckboxConstraint;
  multicheckbox: MultiCheckboxConstraint;
  imagecheckbox: ImageCheckboxConstraint;
  select: SelectConstraint;
  text: TextConstraint;
  MyCustomRoutine: MyCustomRoutineConstraint;
};

type Constraint = ConstraintTypeMap[keyof ConstraintTypeMap] | string[];
